src/memory_region.rs

```diff
@@ -47,10 +49,31 @@ pub struct MemoryRegion {
     pub vm_gap_shift: u8,
     /// Is also writable (otherwise it is readonly)
     pub is_writable: bool,
+    /// Callback executed when a CoW region is first written to.
```

Comment:
Do we really need one callback per region or wouldn't one for all regions in a mapping do? Obviously, that would have to take either the address, reference to or index of the affected region as a parameter.

Reply:
It seems like the least bad choice given the current API to me. Consider:

```
let mem = MemoryRegion::new_cow(ro_slice, addr);
mem.store(42u8, 0x123, 0); // what happens here?
```

We could obviously make it work, but not without making cow memory the "special" sibling memory, where you can use the return value of ::new_readonly and ::new_writable as is, but you can only use cow through MemoryMapping. Doesn't seem great to me.

I have reservations about the current MemoryMapping/MemoryRegion API in general, and I think we need to fix the whole story to among other things properly track lifetimes. In the interest of merging the 80/20 branch (tm), I didn't want to touch the API too much.

Reply:
Ok I actually have an idea that is maaaybe not too ugly, imma give it a try

Reply:
I moved the cow callback up to MemoryMapping. This also required exposing it in EbpfVm::new since the memory mapping instance is created in there